src/sagas/contractSaga.js

```diff
@@ -161,13 +167,50 @@ export function* addToContract(action) {
   }
 }
 
+/**
+ * saga which adds a pasted clause node to the current slate value
+ */
+export function* pasteToContract(action) {
+  try{
+    const { clauseId, clauseTemplateRef, type } = action;
+    const actionRedesign = {
+      clauseId: clauseId,
+      type: type,
+      uri: clauseTemplateRef
+    };
+    // get the templateObj from the store if we already have it or load if not
+    const templateObj = yield call(addTemplateObjectToStore, actionRedesign);
+
+    const metadata = templateObj.getMetadata();
+    const grammar = templateObj.parserManager.getTemplatizedGrammar();
+
+    // Temporary roundtrip and rebuild grammar
+    const grammarRound = roundTrip(grammar);
+    templateObj.parserManager.buildGrammar(grammarRound);
+
+    const sampleText = templateObj.getMetadata().getSamples().default;
+    const model = templateObj.getModelManager().getModels();
+    const logic = templateObj.getScriptManager().getLogic();
+    const clauseTemplateId = uuidv4(); // unique identifier for a clause template
+
+    // add a new clause template to the store so user can edit template
+    yield put(clauseTemplatesActions.addClauseTemplate({
+      metadata, model, logic, sampleText, grammar, id: clauseTemplateId
+    }));
+
+    // add instatiated clause to list of clauses in the contract state
+    yield put(actions.pasteToContractSuccess(clauseId, clauseTemplateId));
```

Comment:
Going back to the question you posed earlier today in Slack, I'm thinking we would want to parse the clause here and send the result/error with the action to be put on the store, since I don't believe we have anything stopping someone from copying and pasting a clause that is in an error state.

Reply:
We can parse with the [`parseClause`](https://github.com/accordproject/template-studio-v2/blob/master/src/utilities/parseClause.js) function here in TSv2.

However, we do something with an annotation in the [`ClausePlugin` in `cicero-ui`](https://github.com/accordproject/cicero-ui/blob/master/src/plugins/ClausePlugin.js#L126)

I'm unsure if this is an issue, this may be where we are indicating to the editor where to color a `parseError`?

Reply:
ah, yes, the annotations is what highlights the text in red. that happens in the `onChange` in the ClausePlugin if the current selection contains a clause. but you are saying that when we copy & paste, the current selection is outside of the clause so we end up calling `next()` [here](https://github.com/accordproject/cicero-ui/blob/44c1354e7cc455597bc9c3513855f455347d3b87/src/plugins/ClausePlugin.js#L236), right? hmm, I wonder if we want to use slate to change the selection to be in the clause after paste. Idk if they would make sense for the user experience though...